src/main/java/com/algolia/search/SearchClient.java

```diff
@@ -0,0 +1,197 @@
+ * This client is intended to be reused and it's thread-safe.
+ *
+ * @see <a href="https://www.algolia.com/doc/rest-api/search/">Algolia.com</a>
+ */
+@SuppressWarnings("WeakerAccess")
+public final class SearchClient
+    implements Closeable,
+        SearchClientMultipleOperations,
+        SearchClientCopyOperations,
+        SearchClientMcm,
+        SearchClientAPIKeys,
+        SearchClientPersonalization,
+        SearchClientAdvanced {
+
+  /** The transport layer. Must be reused. */
+  private final HttpTransport transport;
+
+  /** Client's configuration */
+  private final ConfigBase config;
+
+  /**
+   * Creates a {@link SearchClient} with the given credentials
+   *
+   * @param applicationID The Algolia Application ID
+   * @param apiKey The Algolia API Key
+   * @throws NullPointerException if ApplicationID/ApiKey is null
+   */
+  public SearchClient(@Nonnull String applicationID, @Nonnull String apiKey) {
+    this(new SearchConfig.Builder(applicationID, apiKey).build());
+  }
+
+  /**
+   * Creates a {@link SearchClient} with the given {@link SearchConfig}
+   *
+   * @param config The configuration allows you to advanced configuration of the clients such as
+   *     batch size or custom hosts.
+   * @throws NullPointerException if Config is null
+   */
+  public SearchClient(@Nonnull SearchConfig config) {
+    this(config, new ApacheHttpRequester(config));
+  }
+
+  /**
+   * Creates a {@link SearchClient} with the given {@link SearchConfig}
+   *
+   * @param config The configuration allows you to advanced configuration of the clients such as
+   *     batch size or custom hosts.
+   * @param httpRequester Another HTTP Client than the default one.
+   * @throws NullPointerException if ApplicationID/ApiKey/Config/Requester is null
+   */
+  public SearchClient(@Nonnull SearchConfig config, @Nonnull HttpRequester httpRequester) {
+
+    Objects.requireNonNull(httpRequester, "An httpRequester is required.");
+    Objects.requireNonNull(config, "A configuration is required.");
+
+    this.config = config;
+    this.transport = new HttpTransport(config, httpRequester);
+  }
+
+  /**
+   * Close the underlying Http Client
+   *
+   * @throws IOException if an I/O error occurs
+   */
+  @Override
+  public void close() throws IOException {
+    transport.close();
+  }
+
+  @Override
+  public HttpTransport getTransport() {
+    return transport;
+  }
+
+  @Override
+  public ConfigBase getConfig() {
+    return config;
+  }
+
+  /**
+   * Get the index object initialized (no server call needed for initialization)
+   *
+   * @param indexName The name of the Algolia index
+   * @throws NullPointerException When indexName is null or empty
+   */
+  public SearchIndex<?> initIndex(@Nonnull String indexName) {
+
+    Objects.requireNonNull(indexName, "indexName can't be null.");
+
+    if (AlgoliaUtils.isEmptyWhiteSpace(indexName)) {
+      throw new NullPointerException("The index name is required");
```

Comment:
On these cases I would throw an `InvalidArgumentException`, as the test does not fail only when the String is null

Reply:
👍 I was not sure about this one. thanks

Reply:
Fixed with 8bf447d